packages/builder-vite/plugins/mdx-plugin.ts

```diff
@@ -0,0 +1,71 @@
+import type { Options } from '@storybook/core-common';
+import { Plugin, transformWithEsbuild } from 'vite';
+
+const isStorybookMdx = (id: string) => id.endsWith('stories.mdx') || id.endsWith('story.mdx');
+
+/**
+ * Storybook uses two different loaders when dealing with MDX:
+ *
+ * - *stories.mdx and *story.mdx are compiled with the CSF compiler
+ * - *.mdx are compiled with the MDX compiler directly
+ *
+ * @see https://github.com/storybookjs/storybook/blob/next/addons/docs/docs/recipes.md#csf-stories-with-arbitrary-mdx
+ */
+export function mdxPlugin(options: Options): Plugin {
+  const { features } = options;
+
+  if (features?.previewMdx2) {
+    return {
+      name: 'storybook-vite-mdx2-plugin',
+      enforce: 'pre',
+      async transform(code, id, ssr) {
+        if (id.endsWith('.mdx')) {
+          // @ts-ignore
+          const { compile } = await import('@storybook/mdx2-csf');
+          let mdxCode = String(await compile(code, { skipCsf: !isStorybookMdx(id) }));
+
+          if (isStorybookMdx(id)) {
+            const transformResult = (
+              await transformWithEsbuild(mdxCode, id, {
```

Comment:
I thought it was better to transform with esbuild than try and get the react plugin like vite-plugin-mdx https://github.com/brillout/vite-plugin-mdx/blob/ee4a732f19bd8ee072dc5257b96c3012f2cd6d75/src/index.ts#L45